redisgears_v8_plugin/src/v8_native_functions.rs

```diff
@@ -492,11 +495,14 @@ impl<'isolate_scope, 'isolate, 'ctx_scope, 'a>
 fn add_call_function(
     ctx_scope: &V8ContextScope,
     redis_client: &Arc<RefCell<RedisClient>>,
+    script_ctx: &Arc<V8ScriptCtx>,
```

Comment:
Not following this comment, we need &Arc because the function take a weak ref on it.

Reply:
We create a weak reference of the second line of this function and the closure takes ownership on it.

Reply:
If this could happened then there is a big issue with rust borrow checker. If I have a reference to Arc then there must be an actual Arc behind it so its refcount can not get to zero. And I can not hold the &Arc for a longer time than the Arc itself, the borrow checker will verify it.

Reply:
> Rust borrow checker checks only this Arc reference. It doesn't check for the contents. You can hold a reference to one Arc, while there is another Arc which mutates the contents of this same Arc.

Arc gives you a none mutable reference on, if you want to mutate the object you need a mutable reference which means that you need to also use mutex or something else that will synchronise the writes.

> Regarding the lifetimes - yes, it solves the issue perfectly since we will still have to have the original, owned Arc we refer to. But this also doesn't make sense since we don't need to use the reference, for it will be just used as an owned object (by cloning it later, as you said). That means we can already pass it by "value" (by "move"), and we won't increase the counters even if we pass it by value (by move) instead of a reference as of now. So there are no downsides to passing it by move rather than by reference even so.

Passing ownership of the Arc without cloning will not allows us to use it after and we do need to use it after. 

This is an inner function that gets the &Arc to register some API that needs it. We could have done it without an inner function but for code organisation and readability it is much better this way. We do not violate any safety guarantees and its totally safe to do it so I honestly do not see why not.

Reply:
> So the Arc object holds a pointer to the ArcInner, which has two atomic counters, which are incremeneted/decremented in the RAII-way (new/clone/drop). By passing a reference to Arc - &Arc, what we effectively do is allow ourselves only to use methods of it which receive &self, so const methods. This doesn't mean other objects sharing the same pointer won't increase/decrease the counters inside.

Your example uses pointer which is unsafe from the first place.. if you use unsafe code you need to make sure to wrap it with a safe API and if you do then you should not be able to do what you are describing ...